smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,251 @@
+package software.amazon.smithy.swift.codegen
+
+import software.amazon.smithy.aws.traits.ServiceTrait
+import software.amazon.smithy.aws.traits.auth.SigV4Trait
+import software.amazon.smithy.aws.traits.auth.UnsignedPayloadTrait
+import software.amazon.smithy.model.knowledge.ServiceIndex
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ShapeId
+import software.amazon.smithy.model.traits.AuthTrait
+import software.amazon.smithy.model.traits.OptionalAuthTrait
+import software.amazon.smithy.model.traits.Trait
+import software.amazon.smithy.swift.codegen.integration.ProtocolGenerator
+import software.amazon.smithy.swift.codegen.integration.ServiceTypes
+import software.amazon.smithy.swift.codegen.utils.clientName
+import software.amazon.smithy.swift.codegen.utils.toLowerCamelCase
+
+const val AUTH_SCHEME_RESOLVER = "AuthSchemeResolver"
+const val AUTH_RESOLVER_PARAMS = "AuthSchemeResolverParameters"
+
+class AuthSchemeResolverGenerator() {
+    fun render(ctx: ProtocolGenerator.GenerationContext) {
+        val rootNamespace = ctx.settings.moduleName
+        val serviceIndex = ServiceIndex(ctx.model)
+
+        ctx.delegator.useFileWriter("./$rootNamespace/$AUTH_SCHEME_RESOLVER.swift") {
+            renderResolverParams(serviceIndex, ctx, it)
+            it.write("")
+            renderResolverProtocol(ctx, it)
+            it.write("")
+            renderDefaultResolver(serviceIndex, ctx, it)
+            it.write("")
+            it.addImport(SwiftDependency.CLIENT_RUNTIME.target)
+            it.addImport("AWSClientRuntime")
+        }
+    }
+
+    private fun renderResolverParams(
+        serviceIndex: ServiceIndex,
+        ctx: ProtocolGenerator.GenerationContext,
+        writer: SwiftWriter
+    ) {
+        writer.apply {
+            openBlock(
+                "public struct ${getSdkId(ctx)}$AUTH_RESOLVER_PARAMS: \$L {",
+                "}",
+                ServiceTypes.AuthSchemeResolverParams
+            ) {
+                write("public let operation: String")
+                // If service supports SigV4 auth scheme, it's a special-case
+                // Region has to be in params in addition to operation string from AuthSchemeResolver protocol
+                if (serviceIndex.getEffectiveAuthSchemes(ctx.service).contains(SigV4Trait.ID)) {
+                    write("// Region is used for SigV4 auth scheme")
+                    write("public let region: String?")
+                }
+            }
+        }
+    }
+
+    private fun renderResolverProtocol(ctx: ProtocolGenerator.GenerationContext, writer: SwiftWriter) {
+        writer.apply {
+            openBlock(
+                "public protocol ${getSdkId(ctx)}$AUTH_SCHEME_RESOLVER: \$L {",
+                "}",
+                ServiceTypes.AuthSchemeResolver
+            ) {
+                write("// Intentionally empty.")
```

Comment:
what's the intention of this? Want to make it clear for those reading this code why it is intentionally empty

Reply:
This is just a parent protocol that all auth scheme resolver implementations of a given service must conform to. Added the comment to clarify.